src/psij/executors/batch/batch_scheduler_executor.py

```diff
@@ -0,0 +1,539 @@
+        }
+
+    The *script directory* is a directory (typically `~/.psij/`) where submit scripts are written;
+    it is also used for auxiliary files, such as the *exit code file* (see below) or the *script
+    output file*.
+
+    The *launch command* is a list of strings which the script generator should render as the
+    command to execute. It wraps the job executable in the proper :class:`~Launcher`.
+
+    The function library is a dictionary mapping function names to functions for all public
+    functions in the :mod:`~template_function_library` module.
+
+    The submit script *must* perform two essential actions:
+        1. redirect the output of the executable part of the script to the *script output file*,
+        which is a file in `<script_dir>` named `<native_id>.out`, where `<native_id>` is the id
+        given to the job by the queuing system.
+        2. store the exit code of the launch command in the *exit code file* named
+        `<native_id>.ec`, also inside `<script_dir>`.
+
+    Once the submit script is generated, the executor renders the submit command using
+    :func:`~get_submit_command` and executes it. Its output is then parsed using
+    :func:`~job_id_from_submit_output` to retrieve the `native_id` of the job. Subsequently, the
+    job is registered with the queue polling thread.
+
+    The queue polling thread regularly polls the batch scheduler queue for updates to job states.
+    It builds the command for polling the queue using :func:`~get_status_command`, which takes a
+    list of `native_id` strings corresponding to all registered jobs. Implementations are
+    strongly encouraged to restrict the query of job states to the specified jobs in order to reduce
+    the load on the queuing system. The output of the status command is then parsed using
+    :func:`~parse_status_output` and the status of each job is updated accordingly. If the status
+    of a registered job is not found in the ouput of the queue status command, it is assumed
+    completed (or failed, depending on its exit code), since most queuing systems automatically
+    purge completed jobs from their databases after a short period of time. The exit code is read
+    from the exit code file, as described above. If the exit code value is not zero, the job is
+    assumed failed and an attempt is made to read an error message from the *script output file*.
+    """
+
+    def __init__(self, url: Optional[str] = None,
+                 config: Optional[BatchSchedulerExecutorConfig] = None):
+        """Initializes a `BatchSchedulerExecutor`.
+
+        Parameters
+        ----------
+        url
+            An optional URL pointing to a specific backend
+        config
+            An configuration for this executor instance; if none is specified, a default
+            configuration is used.
+        """
+        super().__init__(url=url, config=config if config else BatchSchedulerExecutorConfig())
+        assert config
```

Comment:
Concrete executors must ensure that a valid config is always passed to the base class. However, the base class cannot instantiate a configuration that is specific to a concrete class. An alternative might be an abstract method that instantiates a concrete config. I've added some explanation to JobExecutor, but the reason for the Optional config is that there are some contexts (the from-string instantiation) where mypy checks against the signature of the base class when instantiating a concrete class and we want to allow a no-config instantiation.

Reply:
Notabug

Reply:
I am sorry, I probably wasn't clear, and probably also missing something, which is a bad combination...  But please indulge me.

To me the above reads like:
```py
def foo(bar : Optional[int] = None):
    assert(bar)
```

which seems to always trigger on `foo()`.

You said: `Concrete executors must ensure that a valid config is always passed to the base class.`.  Well, this is a concrete c'tor, and I understand that you make sure that even if no config is given you pass a valid config to the base class c'tor, that's fine.  But how does that change the fact that the `assert` will raise when a None-config was given (which is the default value)?  Or in other words, If `config` is asserted, why is the default `None`?  I don't think the base c'tor call plays into it, as it won't change the local variable `config`.

Reply:
Config is asserted because there must always be a valid concrete config passed to this constructor.

I want to catch situations when valid concrete configs are not passed by concrete executors, so I set the default to some known value that I can easily test for, hence both the default of None and the assert.

I would like to do that statically, by not having the config optional at all, but then, my factory instantiation will fail without a config, since it's typechecked against the base class, even if my concrete executor may not require a config. In other words, if I change it to a non-optional, instantiation in JobExecutor:214 fails to pass typecheck because, there, and in all concrete classes, I want config to be optional.

So `Optional[]` says "it's ok to construct a sublcass of JobExecutor without a config" and the `assert not None` says "but all of those subclasses must ensure that they substitute a default if they get None for a config".